run

```diff
@@ -1,22 +1,91 @@
-#!/bin/sh
+#!/usr/bin/with-contenv bashio
 
-set -e
+USER="homeassistant"
+PUID="${PUID:-1000}"
+PGID="${PGID:-1000}"
 
-VENV="${VENV:-/var/tmp/venv}"
-log(){ echo "$*" >&2; }
+UMASK="${UMASK:-}"
 
-log "Running as $(id -u):$(id -g)"
+PACKAGES="${PACKAGES:-}"
 
-if [ -n "$UMASK" ]; then
-  log "Setting umask to $UMASK"
+VENV_PATH="${VENV:-/var/tmp/homeassistant-venv}"
+CONFIG_PATH=/config
+
+# Enable Jemalloc for Home Assistant Core
+export LD_PRELOAD="/usr/local/lib/libjemalloc.so.2"
+
+#
+# Creating user
+#
+
+# Some HA commands seem to fail if we don't have an actual user.
+# ie: shell_command would return error code 255
+bashio::log.info "Creating user $USER with $PUID:$PGID"
+
+deluser "$USER" >/dev/null 2>&1 || true
+delgroup "$USER" >/dev/null 2>&1 || true
+
+addgroup -g "$PGID" "$USER"
+adduser -G "$USER" -D -H -u "$PUID" "$USER"
+
+if [ -n "${EXTRA_GID:-}" ]; then
+  bashio::log.info "Resolving suppementary GIDs: $EXTRA_GID"
+  supplementary_groups=()
+
+  for gid in $EXTRA_GID; do
+    group="$(getent group "$gid" | cut -d: -f1 || true)"
+
+    if [ -z "$group" ]; then
+      group="$USER-$gid"
+      addgroup -g "$gid" "$group"
+    fi
+
+    supplementary_groups+=( "$group" )
+  done
+
+  bashio::log.info "Appending supplementary groups: ${supplementary_groups[@]}"
+  for group in "${supplementary_groups[@]}"; do
+    addgroup "$USER" "$group"
+  done
+fi
+
+#
+# Install extra packages
+#
+
+if [ -n "${PACKAGES}" ]; then
+  bashio::log.info "Installing extra packages: $PACKAGES"
+  apk add --quiet --no-progress --no-cache $PACKAGES
+fi
+
+#
+# Create virtual environment
+#
+
+bashio::log.info "Initializing venv in $VENV_PATH"
+su "$USER" \
+  -c "python3 -m venv --system-site-packages '$VENV_PATH'"
+
+#
+# Fix permissions
+#
+
+cd "$CONFIG_PATH" || bashio::exit.nok "Can't find config folder: $CONFIG_PATH"
+chown -R "$USER:$USER" .
+
+if [ -n "${UMASK}" ]; then
+  bashio::log.info "Setting umask: $UMASK"
   umask "$UMASK"
 fi
 
-log "Initializing venv in $VENV"
-python3 -m venv --system-site-packages "$VENV"
-. "$VENV/bin/activate"
+#
+# Run homeassistant
+#
+
+bashio::log.info "Activating venv"
+. "$VENV_PATH/bin/activate"
 
-log "Starting homeassistant"
-set -- ${@:-python3 -m homeassistant --config /config}
-# replace the current pid 1 with the original command
-exec "$@"
+bashio::log.info "Starting homeassistant"
+exec \
+  /bin/s6-setuidgid "$USER" \
+  /usr/local/bin/python3 -m homeassistant --config "$CONFIG_PATH"
```

Comment:
This invokes the system `python3` rather than the `python3` in the `bin` directory of the venv (`/var/tmp/homeassistant-venv/bin/python3` by default).

If `python3` is not found on the `PATH` after activating the venv, there's probably an issue with `virtualenv`. Maybe a `hash -r` is also necessary?

Reply:
Good point, reverted that to python3.

@Nephiel, I cannot reproduce your path issue so maybe you want to try the `hash -r` on your side?

Reply:
No go. This is the run script, only adding the `hash -r` and log the `PATH`:
```
#!/bin/sh

set -e

VENV="${VENV:-/var/tmp/venv}"
log(){ echo "$*" >&2; }

log "Running as $(id -u):$(id -g)"

if [ -n "$UMASK" ]; then
  log "Setting umask to $UMASK"
  umask "$UMASK"
fi

log "Initializing venv in $VENV"
hash -r
log "PATH is $PATH"
python3 -m venv --system-site-packages "$VENV"
. "$VENV/bin/activate"

log "Starting homeassistant"
set -- ${@:-python3 -m homeassistant --config /config}
# replace the current pid 1 with the original command
exec "$@"
```
Log reveals that `PATH` does not contain `/usr/local/bin` at that point:
```
[s6-init] making user provided files available at /var/run/s6/etc...exited 0.
[s6-init] ensuring user provided files have correct perms...exited 0.
[fix-attrs.d] applying ownership & permissions fixes...
[fix-attrs.d] done.
[cont-init.d] executing container initialization scripts...
[cont-init.d] udev.sh: executing... 
starting version 3.2.9
[20:29:01] INFO: Update udev information
[cont-init.d] udev.sh: exited 0.
[cont-init.d] done.
[services.d] starting services
Running as 0:0
Initializing venv in /var/tmp/venv
PATH is /usr/bin:/usr/sbin:/bin:/sbin
./run: line 18: python3: not found
[services.d] done.
[cont-finish.d] executing container finish scripts...
[cont-finish.d] done.
[s6-finish] waiting for services.
[s6-finish] sending all processes the TERM signal.
[s6-finish] sending all processes the KILL signal and exiting.
```
If I replace `python3` with `/usr/local/bin/python3` in the run script, then the container starts fine and I can open a shell into the container to check that the PATH there is correct:
```
$ sudo docker exec -it homeassistant sh
/config # echo $PATH
/usr/local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
```

I'm thinking this must be something local to my setup, though.

Reply:
This is the old run script for HA older than 0.106, not the one provided in my branch. Please use the run script available in this PR.

Reply:
@Chris-V Activating the venv modifies the `PATH`. What does the output look like if you move the `echo $PATH` after the `. "$VENV/bin/activate"`?

Reply:
@Chris-V Looks like I was pulling from master and not this PR... Anyway, that was the issue. Now it runs without any changes.

Reply:
@lwfitzgerald it couldn't have reached that line, because python3 was not in the PATH. But it doesn't really matter (see above).